lib/posix-fdtab/fmap.h

```diff
@@ -0,0 +1,302 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+/* Lock-free data structure specialized in mapping integers to pointers */
+
+#ifndef __UK_FDTAB_FMAP_H__
+#define __UK_FDTAB_FMAP_H__ 1
+
+#include <string.h>
+
+#include <uk/arch/atomic.h>
+#include <uk/assert.h>
+#include <uk/bitops.h>
+#include <uk/bitmap.h>
+#include <uk/thread.h>
+
+/**
+ * Lock-free bitmap, with ones representing a free index.
+ */
+struct uk_bmap {
+	size_t size;
+	volatile unsigned long *bitmap;
+};
+
+#define UK_BMAP_SZ(s) (UK_BITS_TO_LONGS(s) * sizeof(unsigned long))
+
+
+/**
+ * Initialize the bitmap; must not be called concurrently with other functions.
+ */
+static inline void uk_bmap_init(const struct uk_bmap *bm)
+{
+	uk_bitmap_fill((void *)bm->bitmap, bm->size);
+}
+
+/**
+ * Mark `idx` as used, and return whether we were the ones to do so.
+ *
+ * @return
+ * 	0 if we marked `idx` as used, non-zero otherwise
+ */
+static inline int uk_bmap_reserve(const struct uk_bmap *bm, size_t idx)
+{
+	unsigned long mask;
+	unsigned long v;
+
+	if (idx >= bm->size)
+		return -1;
+	mask = UK_BIT_MASK(idx);
+	v = ukarch_and(&bm->bitmap[UK_BIT_WORD(idx)], ~mask);
+	return !(v & mask);
+}
+
+/**
+ * Mark `idx` as free, and return whether we were the ones to do so.
+ *
+ * @return
+ * 	0 if we freed `idx`, non-zero if it was already free or out of range
+ */
+static inline int uk_bmap_free(const struct uk_bmap *bm, size_t idx)
+{
+	unsigned long mask;
+	unsigned long v;
+
+	if (idx >= bm->size)
+		return -1;
+	mask = UK_BIT_MASK(idx);
+	v = ukarch_or(&bm->bitmap[UK_BIT_WORD(idx)], mask);
+	return !!(v & mask);
+}
+
+static inline int uk_bmap_isfree(const struct uk_bmap *bm, size_t idx)
+{
+	if (idx >= bm->size)
+		return -1;
+	return uk_test_bit(idx, bm->bitmap);
+}
+
+/**
+ * Allocate and return the smallest free index larger than `min`.
+ *
+ * @return
+ * 	The allocated index or `>= bm->size` if map full.
+ */
+static inline size_t uk_bmap_request(const struct uk_bmap *bm, size_t min)
+{
+	size_t pos;
+
+	do {
+		/* Seems safe to cast away volatility, revisit if problem */
+		pos = uk_find_next_bit((unsigned long *)bm->bitmap,
+				       bm->size, min);
+		if (pos >= bm->size)
+			return bm->size;
+	/* If bit was already cleared, we lost the race, retry */
+	} while (uk_bmap_reserve(bm, pos));
+
+	return pos;
+}
+
+/**
+ * Mapping of integers to pointers.
+ */
+struct uk_fmap {
+	struct uk_bmap bmap;
+	void *volatile *map;
+};
+
+#define UK_FMAP_SZ(s) ((s) * sizeof(void *))
+
+
+#define UK_FMAP_INRANGE(m, i) ((i) < (m)->bmap.size)
```

Comment:
They indeed do, but since fmap is internal to fdtab, we control all callsites and can ensure this invariant (which is done in `_fdtab_get`), preventing redundant checks. That said, I did find places where validation was missing, so maybe moving all validation into fmap is the better call.

Also, I should probably rename these macros to internal ones; names are left-over from an earlier design where I had intended fmap to have a more generic and perhaps public API.

Reply:
Okay, fair enough if they become somewhat internal.